### PR TITLE
Add function to detect -1 in dataset

### PR DESCRIPTION
**Pull request by hpratama**
- Add function to read if there is any -1 in dataset, and update the retinfo if True
- Set no optional args in dgbpy.hdf5.getOutdType function
- Choice of loss function, choose 'DiceLoss' if withunlabeled is True

### CODE REVIEW

File: dgbpy/hdf5.py [hunk 1]
```diff
@@ -129,8 +129,9 @@ def isSeisClass( info ):
 
 def hasUnlabeled( info ):
   if isinstance(info, dict):
```
Comment (arnauddgbes) [inline dgbpy/hdf5.py]: If I am not mistaken, that can all be put on one line:
```
if isinstance(info, dict) if withunlabeleddictstr in info:
  return info[withunlabeleddictstr]
```

File: dgbpy/hdf5.py [hunk 2]
```diff
@@ -661,6 +657,17 @@ def getInfo( filenm, quick ):
     retinfo.update({plfdictstr: odhdf5.getText(info,'Model.Type')})
   if  odhdf5.hasAttr(info,versionstr):
     retinfo.update({versiondictstr: odhdf5.getText(info,versionstr)})
+  if img2img & isclassification:
```
Comment (arnauddgbes) [inline dgbpy/hdf5.py]: please use the 'and' instead of '&'

File: dgbpy/hdf5.py [hunk 3]
```diff
@@ -661,6 +657,17 @@ def getInfo( filenm, quick ):
...
+    groups = retinfo[exampledictstr].keys()
```
Comment (arnauddgbes) [inline dgbpy/hdf5.py]: First try to read the value from the ++info++ section, like before:
```
hasunlabels = False
if odhdf5.hasAttr(info, withunlabeleddictstr):
  hasunlabels = odhdf5.getBoolValue( info, withunlabeleddictstr )
```
Then only do your implementation below as part of the else block if that same if statement

File: dgbpy/mlapply.py [hunk 4]
```diff
@@ -376,9 +376,13 @@ def doTrain( examplefilenm, platform=dgbkeys.kerasplfnm, type=TrainType.New,
                                           nbchunks=params['nbchunk'],
                                           force=False,
                                           split=params['split'],nbfolds=params['nbfold'] )
-      if type != TrainType.New and dgbkeys.trainconfigdictstr in infos:
-        params[dgbkeys.criteriondictstr] = infos[dgbkeys.trainconfigdictstr][dgbkeys.criteriondictstr]
+      if type != TrainType.New and dgbkeys.trainconfigdictstr in infos and dgbkeys.withunlabeleddictstr in dgbmlio.getInfo(examplefilenm):
```
Comment (arnauddgbes) [inline dgbpy/mlapply.py]: use dgbpy.hdf5.hasUnlabeled( infos )
Store it as a local variable, so you do not need to repeat that function call

File: dgbpy/mlapply.py [hunk 5]
```diff
@@ -376,9 +376,13 @@ def doTrain( examplefilenm, platform=dgbkeys.kerasplfnm, type=TrainType.New,
...
+        params[dgbkeys.criteriondictstr] = 'DiceLoss'
       if type == TrainType.New:
+        if dgbkeys.withunlabeleddictstr in dgbmlio.getInfo(examplefilenm):
```
Comment (arnauddgbes) [inline dgbpy/mlapply.py]: Can be combined with the if statement above

**Review by arnauddgbes — CHANGES_REQUESTED** [review]

File: dgbpy/hdf5.py [hunk 6]
```diff
@@ -657,17 +656,21 @@ def getInfo( filenm, quick ):
...
+  else:
+    if img2img and isclassification:
+      groups = retinfo[exampledictstr].keys()
+      ret = list()
+      for groupnm in groups:
+        try:
+            grp = h5file[groupnm]
+            for inpnm in grp:
+              if np.any(np.array(grp[inpnm][ydatadictstr])==-1):
+                retinfo.update({withunlabeleddictstr: True})
```
Comment (arnauddgbes) [inline dgbpy/hdf5.py]: misplaced. Set hasunlabels, and break both for loops

File: dgbpy/hdf5.py [hunk 7]
```diff
@@ -657,17 +656,21 @@ def getInfo( filenm, quick ):
...
+    hasunlabels = odhdf5.getBoolValue( info, withunlabeleddictstr )
+  else:
+    if img2img and isclassification:
+      groups = retinfo[exampledictstr].keys()
+      ret = list()
+      for groupnm in groups:
+        try:
+            grp = h5file[groupnm]
+            for inpnm in grp:
+              if np.any(np.array(grp[inpnm][ydatadictstr])==-1):
```
Comment (arnauddgbes) [inline dgbpy/hdf5.py:672]: Explicitely check the presence of ydatadictstr in grp[inpnm], if missing, continue

File: dgbpy/hdf5.py [hunk 8]
```diff
@@ -657,17 +656,21 @@ def getInfo( filenm, quick ):
...
+                retinfo.update({withunlabeleddictstr: True})
+        except:
+          pass
```
Comment (arnauddgbes) [inline dgbpy/hdf5.py:678]: when you are done with the if / else block, add the key to the retinfo if hasunlabels == True

File: dgbpy/hdf5.py [hunk 9]
```diff
@@ -657,17 +656,21 @@ def getInfo( filenm, quick ):
...
-    ret = list()
-    for groupnm in groups:
-      try:
-          grp = h5file[groupnm]
-          for inpnm in grp:
-            if np.any(np.array(grp[inpnm][ydatadictstr])==-1):
-              retinfo.update({withunlabeleddictstr: True})
-      except:
-        pass
+  hasunlabels = False
```
Comment (arnauddgbes) [inline dgbpy/hdf5.py]: The condition if img2img and isclassification should stay. None of this should be done otherwise

File: dgbpy/mlapply.py [hunk 10]
```diff
@@ -376,10 +376,11 @@ def doTrain( examplefilenm, platform=dgbkeys.kerasplfnm, type=TrainType.New,
                                           nbchunks=params['nbchunk'],
                                           force=False,
                                           split=params['split'],nbfolds=params['nbfold'] )
-      if type != TrainType.New and dgbkeys.trainconfigdictstr in infos and dgbkeys.withunlabeleddictstr in dgbmlio.getInfo(examplefilenm):
+      hasUnlabels = dgbhdf5.hasUnlabeled(dgbmlio.getInfo(examplefilenm))
```
Comment (arnauddgbes) [inline dgbpy/mlapply.py]: You already have dgbmlio.getInfo(examplefilenm), this is the variable infos
Do not use capital letters in variable names

File: dgbpy/mlapply.py [hunk 11]
```diff
@@ -376,10 +376,11 @@ def doTrain( examplefilenm, platform=dgbkeys.kerasplfnm, type=TrainType.New,
...
+      if type != TrainType.New and dgbkeys.trainconfigdictstr in infos and hasUnlabels:
```
Comment (arnauddgbes) [inline dgbpy/mlapply.py]: @marxprop I still think that both if statements below can be combined, am I missing something?

**Review by arnauddgbes — CHANGES_REQUESTED** [review]

File: dgbpy/mlapply.py (hunk 10 above)
Comment (hpratama) [reply dgbpy/mlapply.py]: infos from doTrain function is from modelin, not from examplefilenm, that's why I used new getInfo function.

File: dgbpy/hdf5.py [hunk 12]
```diff
@@ -656,21 +656,26 @@ def getInfo( filenm, quick ):
...
       for groupnm in groups:
         try:
             grp = h5file[groupnm]
             for inpnm in grp:
+              if ydatadictstr not in grp[inpnm]:
+                continue
               if np.any(np.array(grp[inpnm][ydatadictstr])==-1):
-                retinfo.update({withunlabeleddictstr: True})
+                hasunlabels = True
+                break
```
Comment (arnauddgbes) [inline dgbpy/hdf5.py:674]: ok, but you also need to break the first for loop, started on line 666.
For this check directly thereafter if hasunlabels is True (from the previous iteration), and break then

**Review by arnauddgbes — CHANGES_REQUESTED** [review]

**Review by arnauddgbes — APPROVED** [review]

**Pull request merged**